instrumentation/opentelemetry-instrumentation-urllib/src/opentelemetry/instrumentation/urllib/__init__.py

```diff
@@ -0,0 +1,233 @@
+`urllib https://docs.python.org/3/library/urllib.html>`_ library.
+
+Usage
+-----
+
+.. code-block:: python
+
+    from urllib import request
+    from libs.opentelemetry.instrumentation.urllib import URLLibInstrumentor
+
+    # You can optionally pass a custom TracerProvider to
+    URLLibInstrumentor().instrument()
+
+    req = request.Request('https://postman-echo.com/post', method="POST")
+    r = request.urlopen(req)
+
+API
+---
+"""
+
+import functools
+import types
+from urllib.request import (  # pylint: disable=no-name-in-module,import-error
+    OpenerDirector,
+    Request,
+)
+
+from opentelemetry import context, propagators
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.metric import (
+    HTTPMetricRecorder,
+    HTTPMetricType,
+    MetricMixin,
+)
+from opentelemetry.instrumentation.urllib.version import __version__
+from opentelemetry.instrumentation.utils import http_status_to_status_code
+from opentelemetry.trace import SpanKind, get_tracer
+from opentelemetry.trace.status import Status, StatusCode
+
+# A key to a context variable to avoid creating duplicate spans when instrumenting
+_SUPPRESS_URLLIB_INSTRUMENTATION_KEY = "suppress_urllib_instrumentation"
+
+
+# pylint: disable=unused-argument
+# pylint: disable=R0915
+def get_default_span_name(method):
+    """Default implementation for name_callback, returns HTTP {method_name}."""
+    return "HTTP {}".format(method).strip()
+
+
+def _instrument(tracer_provider=None, span_callback=None, name_callback=None):
+    """Enables tracing of all requests calls that go through
+    :code:`urllib.Client._make_request`"""
+
+    opener_open = OpenerDirector.open
+
+    @functools.wraps(opener_open)
+    def instrumented_request(opener, fullurl, data=None, timeout=None):
+
+        if isinstance(fullurl, str):
+            request_ = Request(fullurl, data)
+        else:
+            request_ = fullurl
+
+        def get_or_create_headers():
+            return getattr(request_, "headers", {})
+
+        def call_wrapped():
+            return opener_open(opener, request_, data=data, timeout=timeout)
+
+        return _instrumented_requests_call(
+            opener, request_, call_wrapped, get_or_create_headers
+        )
+
+    def _instrumented_requests_call(
+        opener, request, call_wrapped, get_or_create_headers
+    ):  # pylint: disable=too-many-locals
+        if context.get_value("suppress_instrumentation") or context.get_value(
+            _SUPPRESS_URLLIB_INSTRUMENTATION_KEY
+        ):
+            return call_wrapped()
+
+        method = request.get_method().upper()
+        url = request.full_url
+
+        span_name = ""
+        if name_callback is not None:
+            span_name = name_callback()
+        if not span_name or not isinstance(span_name, str):
+            span_name = get_default_span_name(method)
+
+        recorder = URLLibInstrumentor().metric_recorder
```

Comment:
this feels a bit weird, to assume that a class has the appropriate field instantiated without actually being a method in that class.

Any reason not to move this method into URLLibInstrumentor, or pass a reference to the metric_recorder object that would be used the by the instrumented method?

Reply:
Sorry by I didn't catch the idea, `metric_recorder` is initialized inside main class in [_instrument](https://github.com/moaddib666/opentelemetry-python-contrib/blob/feature/urllib-instrumentation/instrumentation/opentelemetry-instrumentation-urllib/src/opentelemetry/instrumentation/urllib/__init__.py#L223 ) method.  Or you are talking about a specific method like
`get_metric_recorder(self):`  ?

Reply:
I'm referring to the fact that you're instantiating a ne wURLLibInstrumentor class, just to retrieve metric_recorder. This requires a reader to know:

1. URLLibInstrumentor is a singleton, which requires reading more code about instrumentors
2. understanding that this is safe because _instrument will *only* be called by URLLibInstrumentor, which may not be clear to someone who modifies this code later.

It would be much clearer if you put this code in as a method into URLLibInstrumentor because:

1. you wouldn't have to instantiate a new URLLibInstrumentor, you could just reference it via `self._metric_recorder`
2. It's much clearer to a reader that this method is intended to only be used with URLLibInstrumentor.